File: 40.组合总和-ii.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode.cn id=40 lang=java
 *
 * [40] 组合总和 II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        List<Integer> path = new ArrayList<>();
        method(candidates, target, path, res, 0);
        return res;
    }

    public void method(int[] candidates, int target, List<Integer> path, List<List<Integer>> res, int startidx) {
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        int len = candidates.length;
        for (int i = startidx; i < len;) {
            int remain = target - candidates[i];
            if (remain < 0) {
                break;
            }
            path.add(candidates[i]);
            method(candidates, remain, path, res, i + 1);
            path.remove(path.size() - 1);
            while (++i < len && candidates[i] == candidates[i - 1]) {
            }
        }
    }
}
// @lc code=end

File: 40.组合总和-ii.js
/*
 * @lc app=leetcode.cn id=40 lang=javascript
 *
 * [40] 组合总和 II
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
    let res = []
    function compute(start, remain, path) {
        if (remain === 0) {
            res.push([...path])
            return
        }
        for (let i = start; i < candidates.length; i++) {
            let r = remain - candidates[i]
            if (r < 0) {
                return
            }
            if (i > start && candidates[i] === candidates[i - 1]) {
                continue
            }
            path.push(candidates[i])
            compute(i + 1, r, path)
            path.pop(candidates[i])
        }
    }
    candidates.sort((a, b) => a - b)
    compute(0, target, [])
    return res
};
// @lc code=end

console.log(combinationSum2([10, 1, 2, 7, 6, 1, 5], 8))